contrib/codegen/src/database.rs

```diff
@@ -0,0 +1,119 @@
+    /// The type inside the structure: struct MyDb(ThisType).
+    connection_type: Type,
+}
+
+const EXAMPLE: &str = "example: `struct MyDatabase(diesel::SqliteConnection);`";
+const ONLY_ON_STRUCTS_MSG: &str = "`database` attribute can only be used on structs";
+const ONLY_UNNAMED_FIELDS: &str = "`database` attribute can only be applied to structs with \
+    exactly one unnamed field";
+const NO_GENERIC_STRUCTS: &str = "`database` attribute cannot be applied to a struct with a \
+    generic type";
+
+fn parse_invocation(attr: TokenStream, input: TokenStream) -> Result<DatabaseInvocation> {
+    let attr_stream2 = ::proc_macro2::TokenStream::from(attr);
+    let attr_span = attr_stream2.span();
+    let string_lit = ::syn::parse2::<LitStr>(attr_stream2)
+        .map_err(|_| attr_span.error("expected string literal"))?;
+
+    let input = ::syn::parse::<DeriveInput>(input).unwrap();
+    if !input.generics.params.is_empty() {
+        return Err(input.span().error(NO_GENERIC_STRUCTS));
+    }
+
+    let structure = match input.data {
+        Data::Struct(s) => s,
+        _ => return Err(input.span().error(ONLY_ON_STRUCTS_MSG))
+    };
+
+    let inner_type = match structure.fields {
+        Fields::Unnamed(ref fields) if fields.unnamed.len() == 1 => {
+            let first = fields.unnamed.first().expect("checked length");
+            first.value().ty.clone()
+        }
+        _ => return Err(structure.fields.span().error(ONLY_UNNAMED_FIELDS).help(EXAMPLE))
+    };
+
+    Ok(DatabaseInvocation {
+        type_name: input.ident,
+        db_name: string_lit.value(),
+        structure: structure,
+        connection_type: inner_type,
+    })
+}
+
+pub fn database_attr(attr: TokenStream, input: TokenStream) -> Result<TokenStream> {
+    let invocation = parse_invocation(attr, input)?;
+
+    let connection_type = &invocation.connection_type;
+    let database_name = &invocation.db_name;
+    let request_guard_type = &invocation.type_name;
+    let pool_type = Ident::new(&format!("{}Pool", request_guard_type), request_guard_type.span());
+
+    // The Rocket config has to be cloned because otherwise without `#![feature(nll)]`, you end up
+    // with an error about not being able to move Rocket when adding to managed state or returning
+    // an `Err(rocket)` since it's borrowed by the `rocket.config()` method call.
+    let tokens = quote! {
+        pub struct #request_guard_type(pub ::rocket_contrib::databases::r2d2::PooledConnection<<#connection_type as ::rocket_contrib::databases::Poolable>::Manager>);
+        pub struct #pool_type(::rocket_contrib::databases::r2d2::Pool<<#connection_type as ::rocket_contrib::databases::Poolable>::Manager>);
+
+        impl #request_guard_type {
+            pub fn fairing() -> impl ::rocket::fairing::Fairing {
+                use ::rocket_contrib::databases::Poolable;
+
+                ::rocket::fairing::AdHoc::on_attach(|rocket| {
+                    let config = rocket.config().clone();
```

Comment:
Why are we doing this? Don't clone.

Reply:
Ah, yes, I meant to drop a comment about this.

The reason the `let` binding and `clone` is necessary is because `borrwck` doesn't like this method otherwise. If you send in `&rocket.config()` as the second argument to `database_config(...)`, you get an error that the borrow is still alive when we package up `rocket` in the `Result` type.

If you turn on NLL, this error goes away as you'd expect since the borrow would end before the method returns, but because of the way the lifetimes are defined, lexical lifetimes believes the borrow is still valid and thus rejects the code.

I may be somewhat incorrect in portions of the explanation as I only have a rudimentary understanding of why not cloning is rejected by the compiler, so a lot of it is guess work. I can confirm that when I turned on NLL that the issue went away.